### ISSUE TITLE
Add support for Pulsar cluster level auto failover (PIP-121)

### ISSUE THREAD

**Issue by LeoSht** [fsharplang-ru/pulsar-client-dotnet#217]
Please add support for cluster-level auto failover similar to Java client.
Related information:
[Cluster Level Fail Over PIP-121](https://github.com/apache/pulsar/issues/13315)
[Java Client MR](https://github.com/apache/pulsar/pull/13316)

**Comment by Lanayx** [comment]
Hi! I'm not actively using Pulsar and not adding new functionality to the library myself (only doing bug fixes), but always happy to support if someone wants to add smth, so if you need this feature - you can send a PR

**Comment by fjod** [comment]
So far I get to the point:
in java code when the cluster address changed, they do [this](https://github.com/apache/pulsar/blob/b05f4360585316870961436b43a0c50d6ea60898/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ControlledClusterFailover.java#L140C18-L141C49):
```
   pulsarClient.updateServiceUrl(serviceUrl);
   pulsarClient.reloadLookUp();
```
On[ update service url](https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java#L923):
```
 log.info("Updating service URL to {}", serviceUrl);
        conf.setServiceUrl(serviceUrl);
        lookup.updateServiceUrl(serviceUrl);
        cnxPool.closeAllConnections();
```
and on[ reload lookup](https://github.com/apache/pulsar/blob/b05f4360585316870961436b43a0c50d6ea60898/pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java#L1009):
```
 if (conf.getServiceUrl().startsWith("http")) {
            lookup = new HttpLookupService(conf, eventLoopGroup);
        } else {
            lookup = new BinaryProtoLookupService(this, conf.getServiceUrl(), conf.getListenerName(), conf.isUseTls(),
                    externalExecutorProvider.getExecutor());
        }
```

So far I did same things in pulsarClient
```
 backgroundTask {
       let q = { config with ServiceAddresses = addresses } 
       do! connectionPool.CloseAsync()       
       connectionPool <- ConnectionPool(q)
       lookupService <- BinaryLookupService(q, connectionPool)       
       }
```
but all producers are still connected to old address. It looks like in Java code when all connections are closed, producers/consumers will go to new lookup to get new addresses(?), but in F# they do not.

**Comment by Lanayx** [comment]
I think you need to ensure that once connection is dropped consumers/producers should already have new address for reconnection. In your code you make connection drop first, so probably reconnect happens immediately to the same address

**Comment by fjod** [comment]
>  once connection is dropped consumers/producers should already have new address for reconnection

I tried to rewrite like this:
```
 backgroundTask {
       let q = { config with ServiceAddresses = addresses }
       let oldPool = connectionPool
       connectionPool <- ConnectionPool(q)
       lookupService <- BinaryLookupService(q, connectionPool)
       do! oldPool.CloseAsync()      
       }
```

Interesting, if I start demo app + 2 clusters on 6650 and 6651 addresses, and a webapp which tells pulsar to swtich from 6650 to 6651 - it works :) But if I start with 6650, then boot up the app to broadcast new address 6651, it fails to reconnect tp 6651 with inner exception
 ```
Unhandled exception. System.TimeoutException: Could not send message to broker within given timeout
   at <StartupCode$Pulsar-Client>.$ProducerImpl.SendMessage@760.MoveNext() in F:\work\pulsar-client-dotnet\src\Pulsar.Client\Internal\ProducerImpl.fs:line 767
```

So, it seems like if there is an established connection to broker, my code fails to change the address. But if there is no connection yet (address is switched before client is connected to broker), it works ok.

**Comment by Lanayx** [comment]
@fjod I think you can create a PR already so I could try it myself and help

**Comment by DSmith-VERX** [comment]
@Lanayx  Did @fjod  ever get that branch/PR over to you so someone else could finish it up?

**Comment by Lanayx** [comment]
@DSmith-VERX No, he did not

**Comment by RobertIndie** [comment]
During recent discussions around this feature, we noticed that the original design of PIP-121 has some limitations and is not very convenient for users to implement custom failover strategies.

Because of this, we are currently working on a new cluster-level failover design. Here is a draft design for the C++ client:  
https://github.com/apache/pulsar-client-cpp/pull/547

The plan is to first introduce this design in the C++ / Python / Node.js clients, and then bring it to the Java / F# clients through a new PIP.

Given this, it might be better to implement this feature in the F# client after the new design is finalized. WDYT? @Lanayx

**Comment by Lanayx** [comment]
@RobertIndie It's a perfect timing, since I was about to start the implementation. It looks like there are just API changes, but underlying logic will stay the same. My plan is to release the _alpha_ version with the old API (also taking into account the mentioned changes [here](https://github.com/apache/pulsar-client-cpp/pull/541)), so people can start testing the logic and once Java API is finalized - to publish the stable version with the new API.

**Comment by RobertIndie** [comment]
Would the alpha version be expected to introduce breaking changes once the new design is rolled out?

We would prefer to avoid adding too much compatibility code later (for example, marking methods as deprecated and maintaining them for a while).

**Comment by Lanayx** [comment]
Yes, compile-time breaking changes when migrating from alpha to stable version are fine, run-time breaking changes are not fine. Compatibility code won't be required, migration steps could be added to the corresponding PR.
However, if the ETA for the new API is 1-2 weeks, I can wait.

**Comment by RobertIndie** [comment]
>  if the ETA for the new API is 1-2 weeks, I can wait.

Yes, it should be ready within 1–2 weeks.

The new API design on the C++ client side is currently under review:  
https://github.com/apache/pulsar-client-cpp/pull/541

Feel free to take a look and share any feedback on the design there as well :)

**Comment by Lanayx** [comment]
@RobertIndie am I correct that [Controlled failover](https://pulsar.apache.org/docs/next/client-libraries-cluster-level-failover/#controlled-failover), is only supposed to work with Java client? Since the example doesn't looks like something available for other languages
```
{
"serviceUrl": "pulsar+ssl://target:6651",
"tlsTrustCertsFilePath": "/security/ca.cert.pem",
"authPluginClassName":"org.apache.pulsar.client.impl.auth.AuthenticationTls",
"authParamsString": " \"tlsCertFile\": \"/security/client.cert.pem\"
    \"tlsKeyFile\": \"/security/client-pk8.pem\" "
}
```

**Comment by Lanayx** [comment]
@RobertIndie I don't see any active PR or PIP in Java repo regarding API changes, is it postponed?

**Comment by RobertIndie** [comment]
> [@RobertIndie](https://github.com/RobertIndie) I don't see any active PR or PIP in Java repo regarding API changes, is it postponed?

Apologies for missing your comment earlier. I approved your PR and thank you for the great work!

**Comment by Lanayx** [comment]
Auto failover support has been added in 3.15.0, so closing this issue.

@RobertIndie thank you for help
